### PR TITLE
HOSTEDCP-1000: Add user CA(s) to ocm-trusted-ca-bundle ConfigMap

### PR DESCRIPTION
**Pull request by CrystalChun**
[HOSTEDCP-1000](https://issues.redhat.com/browse/HOSTEDCP-1000)
Adds the CA certificates for mirror registries specified by the user to the ocm-trusted-ca-bundle ConfigMap created in
https://github.com/openshift/hypershift/pull/2437 so that these certificates may be used to pull images from these mirror registries.

**What this PR does / why we need it**: 
Users in disconnected environments will host their own image registry. The certificates from these registries are needed in order to pull from them.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #[HOSTEDCP-1000](https://issues.redhat.com//browse/HOSTEDCP-1000)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Review by bryan-cox — APPROVED** [review]
/lgtm

File: cmd/install/install.go [hunk 1]
```diff
@@ -736,3 +768,23 @@ func hyperShiftOperatorManifests(opts Options) ([]crclient.Object, error) {
 
 	return objects, nil
 }
+
+func getImageRegistryCABundle() string {
+	if client, err := util.GetClient(); err == nil {
```
Comment (csrwng) [inline cmd/install/install.go:773]: My preference would be to gate this on a capability (check whether the image.config.openshift.io resource exists in the mgmt cluster) and report error if there's an error, instead of silently ignoring any errors that occur.

Comment (CrystalChun) [reply cmd/install/install.go:750]: Thank you for the review! I've added a capability check [here](https://github.com/openshift/hypershift/pull/2713/files#diff-b105e0fd290117fe4056a1e2270006e9ea7a6025f5b57936dff42d50660e0b9cR492) and have logged the missing capability [here](https://github.com/openshift/hypershift/pull/2713/files#diff-b105e0fd290117fe4056a1e2270006e9ea7a6025f5b57936dff42d50660e0b9cR524). 

How does it look?

File: cmd/install/install.go [hunk 2]
```diff
@@ -480,6 +489,46 @@ func hyperShiftOperatorManifests(opts Options) ([]crclient.Object, error) {
 		objects = append(objects, userCABundleCM)
 	}
 
+	if mgmtClusterCaps.Has(capabilities.CapabilityImage) && mgmtClusterCaps.Has(capabilities.CapabilityProxy) {
+		imageRegistryCABundle, err := getImageRegistryCABundle()
+		if err != nil {
+			log.Log.Error(err, "unable to get additional image registry CA bundle")
```
Comment (bryan-cox) [inline cmd/install/install.go:495]: I think we would want to just return the error here. I don't think we want logging here.

File: cmd/install/install.go [hunk 3]
```diff
@@ -480,6 +489,46 @@ func hyperShiftOperatorManifests(opts Options) ([]crclient.Object, error) {
...
+				ObjectMeta: metav1.ObjectMeta{Name: "cluster"},
+				Spec: configapi.ProxySpec{
+					TrustedCA: configapi.ConfigMapNameReference{
+						Name: imageRegistryCABundle.Name,
+					},
+				},
+			}
+			objects = append(objects, proxy)
+		}
+	} else {
```
Comment (bryan-cox) [inline cmd/install/install.go:527]: I think we could just drop this else.

File: cmd/install/install.go (hunk 2 above)
Comment (CrystalChun) [reply cmd/install/install.go]: Gotcha, modified to return the error here instead of logging!

File: cmd/install/install.go (hunk 3 above)
Comment (CrystalChun) [reply cmd/install/install.go]: Modified to reflect suggestion.

File: cmd/install/install.go [hunk 4]
```diff
@@ -480,6 +488,43 @@ func hyperShiftOperatorManifests(opts Options) ([]crclient.Object, error) {
 		objects = append(objects, userCABundleCM)
 	}
 
+	if mgmtClusterCaps.Has(capabilities.CapabilityImage) && mgmtClusterCaps.Has(capabilities.CapabilityProxy) {
+		imageRegistryCABundle, err := getImageRegistryCABundle()
```
Comment (bryan-cox) [inline cmd/install/install.go:492]: This returns a string but you then assign a ConfigMap pointer on line 501. I would suggest declaring a var for your ConfigMap here and create a new variable to capture the string returned from `getImageRegistryCABundle`.

Comment (CrystalChun) [reply cmd/install/install.go]: Thanks for catching this! Modified it so that they're under different variable names.

**Comment by CrystalChun** [comment]
To reviewers:

Based off feedback from this [slack thread](https://redhat-internal.slack.com/archives/C058TF9K37Z/p1688052373903309), I've modified the PR and broke up the commits into three separate ones to make it easier to review the changes:
1. The first commit simply adds image to the management cluster capabilities
2. The second is the main logic for actually adding the CA bundle to the Proxy (which will then be auto-added to the configmap)
3. The last is running the logic for the CA bundle as an init container in the hypershift-operator deployment

Please let me know if these should be further broken out into different PRs or if they should actually be combined. Thanks again for all the reviews and feedback!

**Review by bryan-cox — APPROVED** [review]
/lgtm

File: hypershift-operator/main.go [hunk 5]
```diff
@@ -434,3 +439,148 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
 	log.Info("starting manager")
 	return mgr.Start(ctx)
 }
+
+func NewSetupCommand() *cobra.Command {
```
Comment (enxebre) [inline hypershift-operator/main.go:443]: This should be NewInitCommand to keep naming consistent

File: hypershift-operator/main.go [hunk 6]
```diff
@@ -434,3 +439,148 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+		return nil, fmt.Errorf("unable to get discovery client: %w", err)
+	}
+	return capabilities.DetectManagementClusterCapabilities(discoveryClient)
+}
+
+// getImageRegistryCABundle retrieves the image registry CAs listed under image.config.openshift.io
+// and merges them into one CA bundle
+func getImageRegistryCABundle(client crclient.Client) (string, error) {
+	img := &configapi.Image{}
+	if err := client.Get(context.Background(), types.NamespacedName{Name: "cluster"}, img); err != nil {
```
Comment (enxebre) [inline hypershift-operator/main.go:566]: this context should be injected in the signature, similar to what we do in NewStartCommand

File: hypershift-operator/main.go [hunk 7]
```diff
@@ -434,3 +439,148 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+	return cmd
+}
+
+func runInit(ctx context.Context, log logr.Logger) error {
+	log.Info("Initializing environment for Hypershift Operator")
+	client, err := util.GetClient()
+	if err != nil {
+		return err
+	}
+	config := ctrl.GetConfigOrDie()
```
Comment (enxebre) [inline hypershift-operator/main.go:472]: why GetConfigOrDie vs returning a proper error here?

File: cmd/install/assets/hypershift_operator.go [hunk 8]
```diff
@@ -496,6 +496,28 @@ func (o HyperShiftOperatorDeployment) Build() *appsv1.Deployment {
...
 					PriorityClassName:  HypershiftOperatorPriortyClass,
 					ServiceAccountName: o.ServiceAccount.Name,
+					InitContainers: []corev1.Container{
+						{
+							Name:            "init-environment",
+							Image:           image,
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Command:         []string{"/usr/bin/hypershift-operator"},
+							Args:            []string{"init"},
+							// needed since hypershift operator runs with anyuuid scc
```
Comment (enxebre) [inline cmd/install/assets/hypershift_operator.go:506]: can you elaborate why we need this here?

File: hypershift-operator/main.go [hunk 9]
```diff
@@ -434,3 +439,148 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return fmt.Errorf("unable to get discovery client: %w", err)
+	}
+	mgmtClusterCaps, err := capabilities.DetectManagementClusterCapabilities(discoveryClient)
+	if err != nil {
+		return fmt.Errorf("unable to detect management cluster capabilities: %w", err)
+	}
+
+	if mgmtClusterCaps.Has(capabilities.CapabilityImage) && mgmtClusterCaps.Has(capabilities.CapabilityProxy) {
```
Comment (enxebre) [inline hypershift-operator/main.go:482]: Instead of having the main logic within a big if block, let's return early if the if criteria is not met.

File: hypershift-operator/main.go [hunk 10]
```diff
@@ -434,3 +439,148 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+// getImageRegistryCABundle retrieves the image registry CAs listed under image.config.openshift.io
+// and merges them into one CA bundle
+func getImageRegistryCABundle(client crclient.Client) (string, error) {
+	img := &configapi.Image{}
+	if err := client.Get(context.Background(), types.NamespacedName{Name: "cluster"}, img); err != nil {
+		return "", err
+	}
+	if img != nil && img.Spec.AdditionalTrustedCA.Name != "" {
+		configmap := &corev1.ConfigMap{}
+		if err := client.Get(context.Background(), types.NamespacedName{Name: img.Spec.AdditionalTrustedCA.Name, Namespace: "openshift-config"}, configmap); err != nil {
```
Comment (enxebre) [inline hypershift-operator/main.go:571]: same comment for the context

File: hypershift-operator/main.go [hunk 11]
```diff
@@ -434,3 +439,148 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+					"ca-bundle.crt": imageRegistryCABundle,
+				},
+			}
+			if err := apply(ctx, client, imageRegistryCABundleCM); err != nil {
+				return err
+			}
+
+			// Adding the ConfigMap containing the CA bundle to the Proxy object
+			// will allow the openshift-network-operator to automatically
+			// inject this bundle into the openshift-config-managed-trusted-ca-bundle ConfigMap (generated
```
Comment (enxebre) [inline hypershift-operator/main.go:511]: >// inject this bundle into the openshift-config-managed-trusted-ca-bundle ConfigMap (generated
			// during hypershift install command)

Can you link to where install is doing this?

File: hypershift-operator/main.go [hunk 12]
```diff
@@ -434,3 +439,148 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+			// during hypershift install command)
+			// Reference: https://docs.openshift.com/container-platform/4.12/security/certificates/updating-ca-bundle.html
+			proxy := &configapi.Proxy{
+				TypeMeta: metav1.TypeMeta{
```
Comment (enxebre) [inline hypershift-operator/main.go:515]: Is this intentionally changing the proxy resource of the management cluster?

**Comment by enxebre** [comment]
why is this an init container vs and explicit part of the install command? The way it's right now it will be run every time the operator is run/restarted in every environment, is that what we want?

**Comment by CrystalChun** [comment]
> why is this an init container vs and explicit part of the install command? 

Thank you for the reviews! It was decided to be an init container as opposed to an install command because these objects are resolved at runtime. Do let me know if you believe otherwise!

See slack thread discussion [here](https://redhat-internal.slack.com/archives/C058TF9K37Z/p1688139835515419?thread_ts=1688052373.903309&cid=C058TF9K37Z)

> The way it's right now it will be run every time the operator is run/restarted in every environment, is that what we want?

I think this is ideal in case the user happens to use a different secret with their CA, it would be good to capture this case. Even if the user doesn't change it, then redoing this doesn't cause significant changes. Please let me know if my assumptions are incorrect here!

File: hypershift-operator/main.go (hunk 11 above)
Comment (CrystalChun) [reply hypershift-operator/main.go:511]: https://github.com/openshift/hypershift/blob/80dcc89840817fe96b9b9c738b5707eef8026248/cmd/install/install.go#L483-L490

File: hypershift-operator/main.go (hunk 9 above)
Comment (CrystalChun) [reply hypershift-operator/main.go]: Gotcha, thanks for the suggestion. Modified to reflect this. How does it look?

File: hypershift-operator/main.go (hunk 7 above)
Comment (CrystalChun) [reply hypershift-operator/main.go]: Good catch, thanks! Modified to return a proper error here instead.

File: hypershift-operator/main.go (hunk 6 above)
Comment (CrystalChun) [reply hypershift-operator/main.go]: Thanks! Modified to use the passed in context instead

File: hypershift-operator/main.go (hunk 5 above)
Comment (CrystalChun) [reply hypershift-operator/main.go]: Thanks for letting me know. Using `NewInitCommand` as the function signature now.

File: cmd/install/assets/hypershift_operator.go (hunk 8 above)
Comment (CrystalChun) [reply cmd/install/assets/hypershift_operator.go]: Ah sorry, I didn't take time to understand it or verify. It looks like the hypershift operator runs with scc `nonroot-v2` which requires only the UID to be set on container manifests ([src](https://docs.openshift.com/container-platform/4.12/authentication/managing-security-context-constraints.html#default-sccs_configuring-internal-oauth)). 

```
$ oc describe po -n hypershift operator-fcbf86449-5hhpq
Name:                 operator-fcbf86449-5hhpq
Namespace:            hypershift
Labels:               app=operator
...
                      openshift.io/scc: nonroot-v2
```

I've removed the rest of the security context for this container and have only left the uid. 

Follow-up question:
1. Do you know where/how the `scc` is set for this deployment?

**Review by bryan-cox — APPROVED** [review]
/lgtm

File: hypershift-operator/main.go (hunk 11 above)
Comment (CrystalChun) [reply hypershift-operator/main.go]: Sorry I may have misunderstood - did you mean to put the link to it as a code comment?

**Review by jparrill — CHANGES_REQUESTED** [review]

File: cmd/install/assets/hypershift_operator.go [hunk 13]
```diff
@@ -584,9 +598,10 @@ func (o HyperShiftOperatorDeployment) Build() *appsv1.Deployment {
 	}
 
 	if o.OpenShiftTrustBundle != nil {
-		deployment.Spec.Template.Spec.Containers[0].VolumeMounts = append(deployment.Spec.Template.Spec.Containers[0].VolumeMounts, corev1.VolumeMount{
+		optionalTrustBundle := true
+		deployment.Spec.Template.Spec.InitContainers[0].VolumeMounts = append(deployment.Spec.Template.Spec.InitContainers[0].VolumeMounts, corev1.VolumeMount{
```
[the remark this answers is not in the document]
Comment (bryan-cox) [reply cmd/install/assets/hypershift_operator.go:602]: I think we already did in https://github.com/openshift/hypershift/pull/2437 unless I'm mistaken. Crystal's PR was to also pull in any use defined CA's into that same data bundle as well.

[the remark this answers is not in the document]
Comment (CrystalChun) [reply cmd/install/assets/hypershift_operator.go:602]: Resolved in [slack thread](https://redhat-internal.slack.com/archives/C058TF9K37Z/p1689664459811749)

File: hypershift-operator/main.go [hunk 14]
```diff
@@ -61,6 +66,10 @@ import (
 )
 
 func main() {
+	ctrl.SetLogger(zap.New(zap.UseDevMode(true), zap.JSONEncoder(func(o *zapcore.EncoderConfig) {
```
Comment (enxebre) [inline hypershift-operator/main.go:69]: why is this moved and them duplicated into the new command?

File: hypershift-operator/main.go [hunk 15]
```diff
@@ -438,3 +444,127 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
 }
+
+func NewInitCommand() *cobra.Command {
+	ctrl.SetLogger(zap.New(zap.UseDevMode(true), zap.JSONEncoder(func(o *zapcore.EncoderConfig) {
+		o.EncodeTime = zapcore.RFC3339TimeEncoder
+	})))
+
+	cmd := &cobra.Command{
+		Use:   "init",
+		Short: "Initializes the environment for the Hypershift Operator",
```
Comment (enxebre) [inline hypershift-operator/main.go:455]: Can this description be more specific? As a neutral reader I'm not sure what is "Initializes the environment"

File: cmd/install/assets/hypershift_operator.go [hunk 16]
```diff
@@ -584,9 +598,10 @@ func (o HyperShiftOperatorDeployment) Build() *appsv1.Deployment {
 	}
 
 	if o.OpenShiftTrustBundle != nil {
-		deployment.Spec.Template.Spec.Containers[0].VolumeMounts = append(deployment.Spec.Template.Spec.Containers[0].VolumeMounts, corev1.VolumeMount{
+		optionalTrustBundle := true
```
Comment (enxebre) [inline cmd/install/assets/hypershift_operator.go:601]: This is conditional `if o.OpenShiftTrustBundle != nil {`
But the install command is always enforcing its creation https://github.com/openshift/hypershift/blob/bb04b0a79621406a3c0b7522e50ee51cce70f85f/cmd/install/install.go#L483-L544

am I missing anything?

File: hypershift-operator/main.go [hunk 17]
```diff
@@ -438,3 +444,127 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+			return fmt.Errorf("unable to get CA bundle for image mirror registries: %w", err)
+		}
+		if imageRegistryCABundle != nil {
+			if _, err := trustedCABundle.Write(imageRegistryCABundle.Bytes()); err != nil {
+				return fmt.Errorf("unable to write image registry CA to buffer: %w", err)
+			}
+		}
+	}
+
+	if err := ioutil.WriteFile("/trust-bundle/tls-ca-bundle.pem", trustedCABundle.Bytes(), 0644); err != nil {
```
Comment (enxebre) [inline hypershift-operator/main.go:523]: is /trust-bundle  path arbitrary or is there something we can document for where is coming from?

File: hypershift-operator/main.go [hunk 18]
```diff
@@ -438,3 +444,127 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+		return nil, fmt.Errorf("unable to get kubernetes config: %w", err)
+	}
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get discovery client: %w", err)
+	}
+	return capabilities.DetectManagementClusterCapabilities(discoveryClient)
+}
+
+// getImageRegistryCABundle retrieves the image registry CAs listed under image.config.openshift.io
```
Comment (enxebre) [inline hypershift-operator/main.go:545]: it'd be nice to unit test this function

**Comment by enxebre** [comment]
Thanks! dropped some feedback, looks good to me otherwise.

File: cmd/install/assets/hypershift_operator.go (hunk 16 above)
Comment (bryan-cox) [reply cmd/install/assets/hypershift_operator.go:601]: We actually need to not enforce its creation. Always creating it and then mounting the volume here is causing issues for IBM.

Comment (bryan-cox) [reply cmd/install/assets/hypershift_operator.go:601]: I was trying to fix that with Cesar in https://github.com/openshift/hypershift/pull/2744 but that PR is not working. Cesar thought we would need to wait for this PR, or I suppose we can try to fix it within this PR.

File: hypershift-operator/main.go (hunk 15 above)
Comment (CrystalChun) [reply hypershift-operator/main.go]: Makes sense, thanks for pointing this out! Modified to include both a short and long description. Does it sound better? If not, do you have any suggestions?

File: hypershift-operator/main.go (hunk 14 above)
Comment (CrystalChun) [reply hypershift-operator/main.go:69]: Accidentally duplicated, thank you for catching! Removed it from the new command.

File: cmd/install/assets/hypershift_operator.go (hunk 16 above)
Comment (CrystalChun) [reply cmd/install/assets/hypershift_operator.go]: With what Bryan has mentioned, I'll leave this conditional here since the `OpenshiftTrustBundle` may not exist in the future. Please let me know or reopen this conversation if that is not the case.

File: hypershift-operator/main.go (hunk 17 above)
Comment (CrystalChun) [reply hypershift-operator/main.go:520]: This is the path mounted to the init container https://github.com/openshift/hypershift/pull/2713/files#diff-4edc0a874ce10f9d4f77594c9ff489c3275cb2adc788826a9ea2f48a64c336efR624

The path mounted is arbitrary, should this be documented that this is the expected path somehow (as a const?) or should I choose a different path that makes more sense?

File: hypershift-operator/main.go [hunk 19]
```diff
@@ -438,3 +444,124 @@ func run(ctx context.Context, opts *StartOptions, log logr.Logger) error {
...
+
+	if err := ioutil.WriteFile("/trust-bundle/tls-ca-bundle.pem", trustedCABundle.Bytes(), 0644); err != nil {
+		return err
+	}
+
+	log.Info("Finished initializing environment for Hypershift Operator")
+	return nil
+}
+
+// getMgmtClusterCapabilities initializes the ManagementClusterCabilities object
```
Comment (enxebre) [inline hypershift-operator/main.go:528]: where is this function consumed?

File: hypershift-operator/main.go (hunk 18 above)
Comment (enxebre) [reply hypershift-operator/main.go:542]: Can we move all the "init" logic into init.go?

**Comment by bryan-cox** [comment]
/hold

Juan might have found an issue relating to this PR.
